### ISSUE TITLE
[bug][arm64be]with the optimize option -O3 and volatile keyword will get wrong result

### ISSUE THREAD

**Issue by MLsheng** [llvm/llvm-project#64517]
compiler: clang15.0.4 arm64be
demo:
```
#include <stdint.h>
#include <stdio.h>

uint32_t j[6] = {0, 1, 2, 3, 4, 5};
uint32_t k;

void left(uint8_t b) { k = j[b & 5]; }

#pragma pack(1)
struct {
    signed a : 18;       // 0xc740
    volatile unsigned b : 24;
    signed c : 27;
    unsigned : 20;
    signed : 10;
    signed : 15;
    unsigned : 13;
} d[] = {{51008}};

int main() {
    left(d[0].b);
    left(d[0].c);
    printf("%X\n", k);
    return 0;
}

```
Turn on the optimize option -O3, with the volatile declear variable b, output is `4`, wrong; without the volatile keyword, output is `0`, right.

#### memery layout of the struct
```
d:
	.byte	49                              // 0x31
	.byte	208                             // 0xd0
	.byte	0                               // 0x0
	.byte	0                               // 0x0
	.byte	0                               // 0x0
	.byte	0                               // 0x0
	.byte	0                               // 0x0
	.byte	0                               // 0x0
	.byte	0                               // 0x0
	.zero	7
	.size	d, 16
```
#### assembly
with volatile keyword:
```
adrp	x8, d
add	x8, x8, :lo12:d
mov	w9, #20
adrp	x0, .L.str
add	x0, x0, :lo12:.L.str
mov	x29, sp
ldp	x8, xzr, [x8]
and	x8, x9, x8, lsr #57
adrp	x9, j
add	x9, x9, :lo12:j
ldr	w1, [x9, x8]
adrp	x8, k
str	w1, [x8, :lo12:k]

```
without volatile keyword:
```
adrp	x8, d+8
mov	w9, #20
adrp	x0, .L.str
add	x0, x0, :lo12:.L.str
mov	x29, sp
ldr	x8, [x8, :lo12:d+8]
and	x8, x9, x8, lsr #57
adrp	x9, j
add	x9, x9, :lo12:j
ldr	w1, [x9, x8]
adrp	x8, k
str	w1, [x8, :lo12:k]

```
I guess that the register of variable c  is wrong with the volatile keyword.

#### get wrong result when running pass (71)
```
BISECT: running pass (61) LoopSimplifyPass on left
BISECT: running pass (62) LCSSAPass on left
BISECT: running pass (63) SimplifyCFGPass on left
BISECT: running pass (64) InstCombinePass on left
BISECT: running pass (65) InlinerPass on (main)
BISECT: running pass (66) InlinerPass on (main)
BISECT: running pass (67) PostOrderFunctionAttrsPass on (main)
BISECT: running pass (68) ArgumentPromotionPass on (main)
BISECT: running pass (69) OpenMPOptCGSCCPass on (main)
BISECT: running pass (70) SROAPass on main
BISECT: running pass (71) EarlyCSEPass on main
BISECT: NOT running pass (72) SpeculativeExecutionPass on main
BISECT: NOT running pass (73) JumpThreadingPass on main
```
```
*** IR Dump Before EarlyCSEPass on main ***
; Function Attrs: nofree nounwind
define dso_local i32 @main() local_unnamed_addr #1 {
  %1 = load volatile i128, ptr @d, align 16
  %2 = lshr i128 %1, 86
  %3 = trunc i128 %2 to i8
  %4 = and i8 %3, 5
  %5 = zext i8 %4 to i64
  %6 = getelementptr inbounds [6 x i32], ptr @j, i64 0, i64 %5
  %7 = load i32, ptr %6, align 4, !tbaa !5
  store i32 %7, ptr @k, align 4, !tbaa !5
  %8 = load i128, ptr @d, align 16
  %9 = lshr i128 %8, 59
  %10 = trunc i128 %9 to i8
  %11 = and i8 %10, 5
  %12 = zext i8 %11 to i64
  %13 = getelementptr inbounds [6 x i32], ptr @j, i64 0, i64 %12
  %14 = load i32, ptr %13, align 4, !tbaa !5
  store i32 %14, ptr @k, align 4, !tbaa !5
  %15 = load i32, ptr @k, align 4, !tbaa !5
  %16 = call i32 (ptr, ...) @printf(ptr noundef nonnull @.str, i32 noundef %15)
  ret i32 0
}

*** IR Dump After EarlyCSEPass on main ***
; Function Attrs: nofree nounwind
define dso_local i32 @main() local_unnamed_addr #1 {
  %1 = load volatile i128, ptr @d, align 16
  %2 = lshr i128 %1, 86
  %3 = trunc i128 %2 to i8
  %4 = and i8 %3, 5
  %5 = zext i8 %4 to i64
  %6 = getelementptr inbounds [6 x i32], ptr @j, i64 0, i64 %5
  %7 = load i32, ptr %6, align 4, !tbaa !5
  store i32 %7, ptr @k, align 4, !tbaa !5
  %8 = lshr i128 %1, 59
  %9 = trunc i128 %8 to i8
  %10 = and i8 %9, 5
  %11 = zext i8 %10 to i64
  %12 = getelementptr inbounds [6 x i32], ptr @j, i64 0, i64 %11
  %13 = load i32, ptr %12, align 4, !tbaa !5
  store i32 %13, ptr @k, align 4, !tbaa !5
  %14 = call i32 (ptr, ...) @printf(ptr noundef nonnull @.str, i32 noundef %13)
  ret i32 0
}

```

**Comment by ostannard** [comment]
@MLsheng, has this been fixed? I see that you've also raised #64580, which looks like it's probably a duplicate of this.

**Comment by MLsheng** [comment]
> @MLsheng, has this been fixed? I see that you've also raised #64580, which looks like it's probably a duplicate of this.

not been fixed, #64580 is a duplicate of this, and it's more succinct. so I closed this Issue.
